scarlet/interpolation.py

```diff
@@ -419,3 +419,52 @@ def sinc2D(y, x):
         2-D sinc evaluated in x and y
     '''
     return np.dot(np.sinc(y), np.sinc(x))
+
+
+def subsample_function(y, x, f, dNy, dNx=None, dy=None, dx=None):
+    """Subsample a function
+    Given the expected pixel grid of a function, subsample that function
+    at a grid subdivided in x by `dNx` and y by `dNy`.
+    """
+    # Use the spacing between x values to define the subsampled regions
+    if dx is None:
+        dx = x[1] - x[0]
+    if dy is None:
+        dy = y[1] - y[0]
+    if dNx is None:
+        dNx = dNy
+    assert dNy % 2 == 0, "dNy must be even, received {0}".format(dNy)
+    assert dNx % 2 == 0, "dNx must be even, received {0}".format(dNx)
+    assert np.all(np.isclose(x[1:]-x[:-1], x[1] - x[0])), "x must have equal spacing"
+    assert np.all(np.isclose(y[1:]-y[:-1], y[1] - y[0])), "y must have equal spacing"
+
+    # Create the subsampled interval and use it to sample `f`
+    _x = np.linspace(x[0]-dx/2, x[-1]+dx/2, len(x)*dNx+1)
+    _y = np.linspace(y[0]-dy/2, y[-1]+dy/2, len(y)*dNy+1)
+    return f(_y, _x), _y, _x
+
+
+def apply_2D_trapezoid_rule(y, x, f, dNy, dNx=None, dy=None, dx=None):
+    """Use the trapezoid rule to integrate over a subsampled function
+    2D implementation of the trapezoid rule.
+    See `apply_trapezoid_rule` for a description, with the difference
+    that `f` is a function `f(y,x)`, where we note the c ++`(y,x)` ordering.
+    """
+    if dx is None:
+        dx = x[1] - x[0]
+    if dy is None:
+        dy = y[1] - y[0]
+    if dNx is None:
+        dNx = dNy
+    z, _y, _x = subsample_function(y, x, f, dNy, dNx, dy, dx)
+
+    # Calculate the volume of each sub region
+    dz = 0.4 * (z[:-1, :-1] + z[1:, :-1] + z[:-1, 1:] + z[1:, 1:])
+    volumes = dx * dy * dz / dNy / dNx
+
+    # Sum up the sub regions around each point to
+    # give it the same shape as the original `(y,x)`
+    _dNx = len(_x) // dNx
+    _dNy = len(_y) // dNy
+    volumes = np.array(np.split(np.array(np.split(volumes, _dNx, axis=1)), _dNy, axis=1)).sum(axis=(2, 3))
+    return volumes
```

Comment:
I realised that in general, trapezoid rule will be more accurate than rect convolutions unless we can do the convolution analytically or we take enough samples of the function. However, I wonder if the case might arise that we get non-analytic PSF that are not integrated, in which case, the rect convolution would apply?

Reply:
Perhaps. If so we can just implement a rect convolution method in `interpolation` and either pass an option to `psf.generate_psf_image` to choose which type of integration to use or create a `psf.generate_psf_image_from_rect`. But I would save doing that until we discover that we need it.

Reply:
I agree. Plus I'm not even sure the case might arise